feat(examples): Add common README.md (#20)
unikraft/catalog-for-maintainers: examples/README.md

# Common Instructions

These are common instructions for example applications.
Each application gets their own directory, used for documenting, setting up, configuring, building, running, testing and debugging the application.
Common information, configuration scripts are collected together in other directories such as `../utils`.

## Instructions for Bincompat Apps

The following are common information for examples of applications running in binary compatibility (i.e. **bincompat**) mode.
Instructions assume you are running commands in the application example subdirectory.

> Application running in binary compatibility mode use [`app-elfloader`](https://github.com/unikraft/app-elfloader) images to load and run native Linux ELFs (*Executable and Linking Format* files).

Reviewer: ```suggestion
Applications running in binary compatibility mode use [`app-elfloader`](https://github.com/unikraft/app-elfloader) images to load and run native Linux ELFs (*Executable and Linking Format* files).
```

> The images are generated from using `../library/base`.

Reviewer: ```suggestion
The images are generated using the setup under `../library/base/`.
```

We call them `base` images or `base` kernels.

The images are also stored in the Unikraft registry and can be pulled from there.

### Directory Contents

A typical directory for a bincompat app contains:

* `Kraftfile`: build / run rules, including pulling the `base` image

> * `Dockerfile`: filesystem, including binary and libraries

Reviewer: ```suggestion
* `Dockerfile`: creates the filesystem, including binary and libraries
```

Follow the specific instructions in the `README.md` file in the application directory for specifics on building and running with KraftKit and on using the application.

### Scripted Run

To make it easier to run application in bincompat mode, use the generated scripts.
Follow the steps:

1. Build the application root filesystem as an initial ramdisk from the `Dockerfile` using:

   ```console
   make initrd
   ```

   This creates the `rootfs/` directory and the `rootfs.cpio` file with the root filesystem:

   ```console
   tree rootfs/
   ```

1. Run the scripts:

   ```console
   ./run-qemu-x86_64.sh
   ./run-fc-x86_64.sh
   ./kraft-run-qemu.sh
   ```

   All scripts start a local instance of the `base` image with the application running in bincompat mode.
   The application filesystem (binary executable, libraries, configuration files) are packed inside the `rootfs.cpio` initrd file.
   There is no Kraft script to run Firecracker, as Kraft doesn't currently support running images with networking support using Firecracker.

1. You can run the `strace` or `debug` kernels with the scripts:

   ```console

>    ./run-qemu-x86_64.sh kernels/base_qemu-x86_64-strace

Reviewer: ```suggestion
   ./run-qemu-x86_64.sh ../../kernels/base_qemu-x86_64-strace
```
